treeherder/perf/models.py

```diff
@@ -227,9 +229,47 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
         return len(older_ids) or max_chunk_size
 
 
+class Counter(models.Model):
+    model = models.CharField(max_length=80, blank=False, null=False, primary_key=True)
+    chunk_size = models.IntegerField(null=False, default=100)
+    count = models.BigIntegerField(null=False)
+
+
+COUNTERS = {}
+COUNTERS_LOCK = Lock()
+
+
+def next_id(model):
+    """
+    :return:  A unique integer key for given model
+    """
+    start = 1
+    model_name = model.__name__
+    with COUNTERS_LOCK:
+        count, maxx = COUNTERS.get(model_name, (start, 0))
+        if count >= maxx:
+            # we require a new block of ids
+            with transaction.atomic():
```

Comment:
I am a bit concerned that a transaction in a lock will deadlock. Django makes it hard to control the commands sent to the database, nevermind the controlling the order of those commands.  I hope that updating the `Counter` will be executed, and not accumulated in a transaction to be sent at a later time.

Reply:
This feels complicated & I still don't know if Django uses a multi threaded and/or multi processing execution model. However, Django already provides support for concurrent access to models. Even at row level.

I think you should rather use [selected_for_update()](https://docs.djangoproject.com/en/3.0/ref/models/querysets/#select-for-update) instead.
[Others](https://stackoverflow.com/questions/1030270/race-conditions-in-django) have used this mechanism as well.

Reply:
Django is definitely run on multiple machines and multiple processes: The biggest users are the log parsers.

Reply:
Regarding the global `COUNTERS` variable... I'm not sure that's the best approach.
It could be read/updated not only by web requests, but by cronjobs also or even by one-off Dyno commands.

Reply:
> Regarding the global COUNTERS variable... I'm not sure that's the best approach.
It could be read/updated not only by web requests, but by cronjobs also or even by one-off Dyno commands.

This is what the `COUNTERS_LOCK` is for:  It ensures that each thread in a process has exclusive accesses the COUNTERS.

> The multi thread lock won't suffice for the cross-process problem.

That is what `transaction.atomic():` is for: Ensuring the `db_counter` is unique among all the processes that access the database.

> On SO, people advice against using global shared variables, as Django has a "shared-nothing" architecture.

I completely agree with the philosophy of the "shared nothing" architecture, unfortunately Django and Treeherder are not using that philosophy:  The `AUTO INCREMENT` columns are shared global variables.  `AUTO INCREMENT` columns should not be used:  They are convenient for small apps, but do not scale.

This PR is an attempt to avoid a large database migration, and replace the `AUTO INCREMENT` feature which provides a dense, non-overlapping, id numbers that fit in `INT32`.  Since `AUTO INCREMENT` is a global variable, it makes sense the replacement will act similar.. 

This code has two levels:  One, to serve chunks of ids to the various machines/processes, and another to coordinate the serving of individual ids to each of the threads.

Reply:
> I think you should rather use selected_for_update() instead.

Thank you for mentioning this, I will look at that.  I am not sure if `selected_for_update()` will work when I must create a new record.

```
from django.db import transaction

entries = Entry.objects.select_for_update().filter(author=request.user)
with transaction.atomic():
    for entry in entries:
        ...
```

There is something pathological about starting a transaction outside the block declaring a transaction.

Reply:
> > Regarding the global COUNTERS variable... I'm not sure that's the best approach.
> It could be read/updated not only by web requests, but by cronjobs also or even by one-off Dyno commands.
> 
> This is what the `COUNTERS_LOCK` is for:  It ensures that each thread in a process has exclusive accesses the COUNTERS.

Indeed, it ensures multiple threads from the **same** process don't step on each other's toes.
But I'm more concerned in multi process scenarios.

If we have 2 separate processes A & B, they'll both have their own COUNTERS object *(initially empty)*.
Even if they both start at the same time & lazy initialize COUNTERS from the database in the same way (such that both COUNTERS are initially the same), that doesn't guarantee the ids generated in A & B won't be duplicates; COUNTERS acts as a cache system at process level, built from scratch; A & B don't know each other's COUNTERS. When A & B will try to set these duplicated ids in the database, there are chances the client code will through errors.

In other words, A will think that "Hey! I can use the ids from this X chunk." B will consider that as well, for the very same X chunk. But when they try to sync that to the central database, one of them will fail.

Reply:
> When A & B will try to set these duplicated ids in the database, there are chances the client code will through errors.

You are correct.  That happens when the isolation level set to `REPEATABLE READ` (or less strict). When prototype was in that mode, occasionally two processes will have the correct timing, and both grab the same block at the same time.

With isolation level set to `SERIALIZABLE`, we get ACID compliance: We can assume only one process can enter the `with transaction.atomic()` section at a time:

* A requires a block of ids
* B requires a block of ids
* A reads the count (`N`) out of the database, and locks the record 
* B tries to reads the same count (`N`) out of the database, but is blocked 
* A increments by 100
* A updates the count in the database (`N+100), and releases lock
* B reads the count (`N+100`) out of the database
* B increments by 100
* B updates the count in the database (`N+200) 

Each process will have a unique block.

Reply:
I think an alternative would be the Django's builtin cache system, but I'm not at all familiar with it. @camd may be of help here.

It may be possible to use the cache to "lock" a set of ids.  But I would much prefer using the database to lock something.  That said, though, this is getting outside my domain knowledge.  :)  My db-fu only goes so deep and you guys are already deeper than that.  :)